Sources/Contentful/ArrayResponse.swift

```diff
@@ -107,4 +111,193 @@ public struct MappedArrayResponse<ItemType>: Array where ItemType: EntryModellab
+    let assets: [Asset]?
+    let entries: [EntryDecodable]?
+
+    private enum CodingKeys: String, CodingKey {
+        case assets     = "Asset"
+        case entries    = "Entry"
+    }
+
+    init(from decoder: Decoder) throws {
+        let container   = try decoder.container(keyedBy: CodingKeys.self)
+
+        assets          = try container.decodeIfPresent([Asset].self, forKey: CodingKeys.assets)
+        // Cache to enable link resolution.
+        if let assets = assets {
+            decoder.linkResolver.cache(assets: assets)
+        }
+
+        // A copy as an array of dictionaries just to extract "sys.type" field.
+        guard let jsonItems = try container.decodeIfPresent(Swift.Array<Any>.self, forKey: .entries) as? [[String: Any]] else {
+            self.entries = nil
+            return
+        }
+        var entriesJSONContainer = try container.nestedUnkeyedContainer(forKey: .entries)
+        var entries: [EntryDecodable] = []
+        let contentTypes = decoder.userInfo[DecoderContext.contentTypesContextKey] as! [ContentTypeId: EntryDecodable.Type]
+
+        while entriesJSONContainer.isAtEnd == false {
+            let contentTypeInfo = try jsonItems.contentTypeInfo(at: entriesJSONContainer.currentIndex)
+
+            // For includes, if the type of this entry isn't defined by the user, we skip serialization.
+            if let type = contentTypes[contentTypeInfo.id] {
+                let entryModellable = try type.popEntryDecodable(from: &entriesJSONContainer)
+                entries.append(entryModellable)
+            }
+        }
+        self.entries = entries
+
+        // Cache to enable link resolution.
+        if let entries = self.entries {
+            decoder.linkResolver.cache(entryDecodables: entries)
+        }
+    }
+}
+
+extension MappedArrayResponse: Decodable {
+
+    public init(from decoder: Decoder) throws {
+        let container   = try decoder.container(keyedBy: CodingKeys.self)
+        skip            = try container.decode(UInt.self, forKey: .skip)
+        total           = try container.decode(UInt.self, forKey: .total)
+        limit           = try container.decode(UInt.self, forKey: .limit)
+
+        // All items and includes.
+        includes        = try container.decodeIfPresent(MappedIncludes.self, forKey: .includes)
+
+        // A copy as an array of dictionaries just to extract "sys.type" field.
+        guard let jsonItems = try container.decode(Swift.Array<Any>.self, forKey: .items) as? [[String: Any]] else {
+            throw SDKError.unparseableJSON(data: nil, errorMessage: "SDK was unable to serialize returned resources")
+        }
+        var entriesJSONContainer = try container.nestedUnkeyedContainer(forKey: .items)
+        var entries: [EntryDecodable] = []
+        let contentTypes = decoder.userInfo[DecoderContext.contentTypesContextKey] as! [ContentTypeId: EntryDecodable.Type]
+
+        while entriesJSONContainer.isAtEnd == false {
```

Comment:
same as previous

Reply:
Some of this code looks to be duplicated, may there be a way to de-duplicate and reuse it?

Reply:
unfortunately, these are different types and i'm just initializing a few mutable and immutable variables. i think the only refactor opportunity would be a helper that returns a tuple, but in fact the handling of data inside this json is different while iterating through the container.